### PR TITLE
Improve Dragon Darts

### PR DESCRIPTION
**Pull request by HoeenCoder**
This moves most of dragon dart's logic to `Pokemon.getMovetargets` by giving a new move target type called `allFoesSmart` (Feel free to bikeshed about the name, I'm not too attatched).

Put this up as a draft since I need to do something about being unable to tell if a move's target was forcibly changed during a `RedirectTarget` event. Looking for feedback/assistance on this.

### CODE REVIEW

File: sim/pokemon.ts [hunk 1]
```diff
@@ -628,6 +628,42 @@ export class Pokemon {
 				this.battle.retargetLastMove(targets[targets.length - 1]);
 			}
 			break;
+		// FIXME client support for this move target (allow user to select a target in doubles)
```
Comment (Zarel) [inline sim/pokemon.ts]: Client PR should happen first.

File: data/moves.js [hunk 2]
```diff
@@ -4116,22 +4116,11 @@ let BattleMovedex = {
...
-				move.spreadModifier = 1;
-				move.target = "allAdjacentFoes";
-				for (const currentTarget of target.side.active) {
-					if (currentTarget.volatiles['protect'] || !currentTarget.runImmunity('Dragon') || currentTarget.fainted) {
-						move.multihit = 2;
-						break;
-					}
-				}
-			}
+			if (target.side.active.length === 2) move.spreadModifier = 1;
```
Comment (Zarel) [inline data/moves.js]:
```suggestion
			move.spreadModifier = 1;
```

File: sim/pokemon.ts [hunk 3]
```diff
@@ -628,6 +628,42 @@ export class Pokemon {
...
+		case 'allFoesSmart':
+			// Check for redirection first
+			const newTarget = this.battle.priorityEvent('RedirectTarget', this, this, this.battle.dex.getActiveMove(move), target);
+			this.battle.debug(`Original Target: ${target.side.name} ${target.species}, New Target: ${newTarget.side.name} ${newTarget.species}`);
```
Comment (Zarel) [inline sim/pokemon.ts]: remember to remove this

File: sim/pokemon.ts [hunk 4]
```diff
@@ -628,6 +628,42 @@ export class Pokemon {
 				this.battle.retargetLastMove(targets[targets.length - 1]);
 			}
 			break;
+		// FIXME client support for this move target (allow user to select a target in doubles)
+		case 'allFoesSmart':
+			// Check for redirection first
+			const newTarget = this.battle.priorityEvent('RedirectTarget', this, this, this.battle.dex.getActiveMove(move), target);
```
Comment (Zarel) [inline sim/pokemon.ts]: Is Dragon Darts even redirectable? I thought no.

Comment (Zarel) [reply sim/pokemon.ts]: Ohhh I see from further on, interesting.

File: sim/pokemon.ts (hunk 1 above)
Comment (Zarel) [reply sim/pokemon.ts]: Does selecting a target even do anything? If no, it shouldn't matter.

**Comment by HoeenCoder** [comment]
Made significant changes

File: data/moves.js (hunk 2 above)
Comment (urkerab) [reply data/moves.js]: Or you could just put the `spreadModifier` directly in the move data?

Comment (Zarel) [reply data/moves.js]: I think this probably shouldn't be considered a spread move at all, but a multihit move that smart-targets every hit.

**Comment by Zarel** [comment]
Wow, we had a Dragon Darts PR a monthish ago and I forgot about it?

**Comment by HoeenCoder** [comment]
Yeah my plan is/was to resume work when I get time, though since its high priority if you want to work on it, be my guest.

File: sim/global-types.ts [hunk 5]
```diff
@@ -250,7 +250,7 @@ interface EventMethods {
 	onNegateImmunity?: ((this: Battle, pokemon: Pokemon, type: string) => boolean | void) | boolean
 	onOverrideAction?: (this: Battle, pokemon: Pokemon, target: Pokemon, move: ActiveMove) => string | void
 	onPrepareHit?: CommonHandlers['ResultSourceMove']
-	onRedirectTarget?: (this: Battle, target: Pokemon, source: Pokemon, source2: Effect, move: ActiveMove) => Pokemon | void
+	onRedirectTarget?: (this: Battle, target: Pokemon, source: Pokemon, source2: Effect, move: ActiveMove) => Pokemon | Pokemon[] | void
```
Comment (Zarel) [inline sim/global-types.ts]: The relayVar is of type `Pokemon`, so this return type will actively cause crashes if someone ever adds a lower-priority `RedirectTarget` handler.

File: data/statuses.js [hunk 6]
```diff
@@ -462,6 +462,31 @@ let BattleStatuses = {
 			return this.chainModify([0x14CD, 0x1000]);
 		},
 	},
+	smarttarget: {
+		name: 'smarttarget',
+		id: 'smarttarget',
+		num: 0,
+		duration: 1,
+		onRedirectTargetPriority: -1,
+		onRedirectTarget(target, source, source2, move) {
```
Comment (Zarel) [inline data/statuses.js]: I'm actually really confused how this handler isn't applying for redirected moves.

**Pull request closed without merging**